arch/x86/adler32_avx.c

```diff
@@ -109,41 +114,22 @@ Z_INTERNAL uint32_t adler32_avx2(uint32_t adler, const unsigned char *buf, size_
                 back down to 32 bit precision later (there is in AVX512) 
             3.) Full width integer multiplications aren't cheap
 
-        We can, however, cast up to 64 bit precision on all 8 integers at once, and do a relatively
-        cheap sequence for horizontal sums. Then, we simply do the integer modulus on the resulting
-        64 bit GPR, on a scalar value
+        We can, however, and do a relatively cheap sequence for horizontal sums. 
+        Then, we simply do the integer modulus on the resulting 64 bit GPR, on a scalar value. It was
+        previously thought that casting to 64 bit precision was needed prior to the horizontal sum, but
+        that is simply not the case, as NMAX is defined as the maximum number of scalar sums that can be
+        performed on the maximum possible inputs before overflow
         */
 
  
-        /* Will translate to nops */
-        __m128i s1lo = _mm256_castsi256_si128(vs1);
-        __m128i s2lo = _mm256_castsi256_si128(vs2);
-
-        /* Requires vextracti128 */
-        __m128i s1hi = _mm256_extracti128_si256(vs1, 1);
-        __m128i s2hi = _mm256_extracti128_si256(vs2, 1);
-        
-        /* Convert up to 64 bit precision to prevent overflow */
-        __m256i s1lo256 = _mm256_cvtepi32_epi64(s1lo);
-        __m256i s1hi256 = _mm256_cvtepi32_epi64(s1hi);
-        __m256i s2lo256 = _mm256_cvtepi32_epi64(s2lo);
-        __m256i s2hi256 = _mm256_cvtepi32_epi64(s2hi);
-
-        /* Sum vectors in existing lanes */
-        __m256i s1_sum = _mm256_add_epi64(s1lo256, s1hi256);
-        __m256i s2_sum = _mm256_add_epi64(s2lo256, s2hi256);
-        
         /* In AVX2-land, this trip through GPRs will probably be unvoidable, as there's no cheap and easy
          * conversion from 64 bit integer to 32 bit. This casting to 32 bit is cheap through GPRs 
          * (just register aliasing), and safe, as our base is significantly smaller than UINT32_MAX */
```

Comment:
Looks like this comment might need to get an update. The cast to 32 bit is removed for example.

Reply:
Yes and no.  The optimization the compiler is performing for that fixed modulus involves casting up to 64 bit precision and shifting the carry bits back into 32 bit.  For that reason we don't do the modulus within the vector register - it's still a chore to do so.  We also don't happen to need to do it to all of them, and since summing into the first vector position and moving it into eax is relatively cheap to begin with, there's definitely no reason to prefer to not make the trip to GPRs.

Arguably, if we could inexpensively do the modulus in vector registers, we could avoid the scalar horizontal sum, I guess.  But yes, the sequence needed to go to 64 bit precision and back to 32 bit is quite high, likely higher than the horizontal sum + move to GPR + move back when scoreboarded.

Reply:
Also even if we could cheaply move between 32 bit to 64 bit back to 32 bit precision again in the vector register - ultimately the thing that would murder all of our cycles would be the multiplication.  Full 64 bit multiplication is very expensive, and even though I think the "low" 64 bits are all that we'd need, that instruction has quite a bit of latency associated with it, and we'd need to do it twice since we'd be split over two 256 bit registers for that precision, with no real hope of ILP.

Reply:
I modified the comment to clarify exactly what I'm referring to when I talk about switching precision.  Let me know if it's still too vague.